pkgs/os-specific/linux/zfs/generic.nix

```
@@ -50,9 +51,37 @@ let
     # If you don't do this your ZFS builds will fail on any non-standard (e.g.
     # clang-built) kernels.
     stdenv' = if kernel == null then stdenv else kernel.stdenv;
+
+    isLTS = kernel.passthru.isLTS or false;
   in
 
-  stdenv'.mkDerivation {
+  lib.warnIf (!(buildKernel -> !isLTS -> allowNonLTS)) ''
+    You are using the ZFS kernel module in combination with a non-LTS kernel:
+
+    ${kernel.modDirVersion}
+
+    Because the development cycles of OpenZFS and the Linux kernel do not line
+    up, historical precedent has show it to be extremely likely that this kernel
+    will go EOL (and will therefore be removed from Nixpkgs) before an OpenZFS
+    release is available that supports the (at that time) latest non-LTS kernel.
+    This is the case for any kernel that is not an LTS release.
+
+    See https://www.kernel.org/ for more information on LTS kernel releases.
+
+    When that happens, you will *no longer be able to update* and will receive
+    *no support* from NixOS/Nixpkgs maintainers.
+
+    Please use the latest LTS kernel (`pkgs.linuxPackages`) instead if you wish
+    to continue to use ZFS in the future in a way that NixOS/Nixpkgs supports.
+
+    The ZFS package can be overridden with `allowNonLTS = true` to hide this
+    warning if you are willing to risk running into the described issue. The
+    NixOS option `boot.zfs.allowNonLTS` also exists for convenience. **Do not
+    create issues regarding to ZFS kernel support if you enable this flag; it is
+    not supported and will not be supported.
```

```suggestion
    You have enabled the ZFS kernel module in combination with a non-LTS kernel, 
    which could prevent you from upgrading for days or weeks due to misalignment
    in upstream release cycles. This configuration may received limited support
    accordingly.

    Please use the latest LTS kernel (`pkgs.linuxPackages`), remove zfs, set the 
    NixOS option `boot.zfs.allowNonLTS` also exists for convenience, or see more 
    options at https://wiki.nixos.org/wiki/ZFS
```

Here's a shorter version that I think is still much too long. Let's keep it short and to the point if we're going to warn users.

I'm still not convinced about this constant warning, and may prefer users opt in explicitly.